### DISCUSSION TITLE
[Dove] Add Hooks To Custom Service Methods

### DISCUSSION THREAD

**Discussion by forgot** [feathersjs/feathers#2386] [Q&A] [answered]
I've have custom service methods working, but now I'm trying to figure out how to attach hooks to them. If I try to add the custom method's name to the hooks file generated by the CLI, I get an error:
`Exception has occurred: Error: 'customMethod' is not a valid hook method`

How should hooks be attached to custom methods?

**Comment by daffl** [comment, marked as answer]
Custom methods only work with the [new hooks](https://github.com/feathersjs/hooks) which can be registered like this:

```js
app.service('myservice').hooks({
  myCustomMethod: [
    async (context, next) => {
      console.log(`Doing things before ${context.method}`);
      await next();
      console.log(`Doing things after ${context.method}`);
    }
  }
});
```

Alternatively you can bring them closer to the service class using this format:

```js
import { hooks } from '@feathersjs/hooks';

export class MyService {
  async find (params) {
    return [];
  }

  async myCustomMethod (data, params) {
    return data;
  }
}

// Hooks for all service methods
hooks(MyService.prototype, [
  authenticate('jwt')
]);

hooks(MyService, {
  find: [
    // find hooks here
  ],
  myCustomMethod: [
    async (context, next) => {
      console.log(`Doing things before ${context.method}`);
      await next();
      console.log(`Doing things after ${context.method}`);
    }
  ]
})
```

Or with TypeScript decorators (`"experimentalDecorators": true`):

```ts
import { HookContext } from '@feathersjs/feathers';
import { hooks, NextFunction } from '@feathersjs/hooks';

@hooks([
  authenticate('jwt'),
  async (context: HookContext, next: NextFunction) => {
    console.log(`Class level hook before ${context.method}`);
    await next();
    console.log(`Class level hook after ${context.method}`);
  }
])
export class MyService {
  @hooks([
    // find hooks here
  ])
  async find (params) {

  }

  @hooks([
    async (context: HookContext, next: NextFunction) => {
      console.log(`Doing things before ${context.method}`);
      await next();
      console.log(`Doing things after ${context.method}`);
    }
  ])
  async myCustomMethod (data, params) {
    return data;
  }
}
```

**Reply by forgot** [reply]
Awesome, thank you again for the detailed explanation! I'd been reading about the new hooks [here](https://github.com/feathersjs/hooks), but didn't realize legacy hooks would no longer work for custom service methods.

**Reply by daffl** [reply]
It's another one of those things that is not fully documented at the moment 😅 (and I also still have to add some wrappers to convert legacy hooks to new ones).

But yes, the legacy hooks are mainly for backwards compatibility which didn't have custom methods.

**Reply by forgot** [reply]
No worries, it's getting there. Can both new and legacy hooks be used at the same time? Or should I convert all my current legacy hooks to the new format?

**Reply by daffl** [reply]
Yes, they can be used at the same time. The order (tested in https://github.com/feathersjs/feathers/blob/dove/packages/feathers/test/hooks/hooks.test.ts#L6-L102) is:

- New hooks - app.hooks
- Legacy hooks - app.hooks before
- New hooks - service all
- New hooks - service method
- Legacy hooks before - service all
- Legacy hooks before - service method
- Legacy hooks after - service method
- Legacy hooks after - service all
- Legacy hooks - app.hooks after
